### PR TITLE
Add go-to definition in metadata

### PR DESCRIPTION
**Pull request by cheerio-pixel**
Reviving the [work](https://github.com/emacs-lsp/lsp-mode/pull/2573) of razzmat, I have reworked his pull request to support this feature.

### CODE REVIEW

**Comment by jcs090218** [comment]
@razzmatazz WDYT? 🤔

File: clients/lsp-csharp.el [hunk 1]
```diff
@@ -334,6 +348,61 @@ using the `textDocument/references' request."
...
+    file-location))
+
+(defun lsp-csharp--uri->path-fn (uri)
+  "Custom implementation of lsp--uri-to-path function to glue omnisharp's metadata uri."
+  (if (string-match-p lsp-csharp--metadata-uri-re uri)
+      (lsp-csharp--omnisharp-metadata-uri-handler uri)
+    (lsp--uri-to-path-1 uri))
+  )
+
+(defun lsp-csharp--environment-fn ()
```
Comment (razzmatazz) [inline clients/lsp-csharp.el:401]: I would suggest renaming this to `lsp-csharp--omnisharp-environment-fn` as there are more than 1 LS that are supported in lsp-mode/csharp

File: clients/lsp-csharp.el [hunk 2]
```diff
@@ -334,6 +348,61 @@ using the `textDocument/references' request."
...
+
+      (with-temp-file metadata-file-location
+        (insert uri))
+
+      (with-temp-file file-location
+        (insert source)))
+
+    file-location))
+
+(defun lsp-csharp--uri->path-fn (uri)
```
Comment (razzmatazz) [inline clients/lsp-csharp.el:394]: Needs rename to `lsp-csharp--omnisharp-uri->path-fn`?

File: clients/lsp-csharp.el [hunk 3]
```diff
@@ -334,6 +348,61 @@ using the `textDocument/references' request."
       (lsp-show-xrefs (lsp--locations-to-xref-items locations-found) nil t)
     (message "No references found")))
 
+(defun lsp-csharp--path->qualified-name (path)
```
Comment (razzmatazz) [inline clients/lsp-csharp.el:351]: `lsp-csharp--omnisharp-path->qualified-name`?

File: clients/lsp-csharp.el [hunk 4]
```diff
@@ -41,6 +41,13 @@ Version 1.34.3 minimum is required."
   :link '(url-link "https://github.com/OmniSharp/omnisharp-roslyn")
   :package-version '(lsp-mode . "9.0.0"))
 
+(defconst lsp-csharp--metadata-uri-re
```
Comment (razzmatazz) [inline clients/lsp-csharp.el:44]: `lsp-csharp--omnisharp-metadata-uri-re` ?

**Review by razzmatazz — APPROVED** [review]
Approved with some suggestions to rename omnisharp-related functions to have `omnisharp` in the name.

**Comment by razzmatazz** [comment]
> @razzmatazz WDYT? 🤔

Reviewed, thank you! Approved 👍🏻 , with a suggestion to rename omnisharp-related fns to have `omnisharp` in the name

**Comment by jcs090218** [comment]
@cheerio-pixel Friendly ping! :)

**Comment by cheerio-pixel** [comment]
> @cheerio-pixel Friendly ping! :)

Sorry, completly forgot everything about this. Thank for the ping.

**Comment by cheerio-pixel** [comment]
@jcs090218 Finished with the recommendations. Sorry for messing up with the historial. I didn't realize that rebase modifies the authoring information.

**Comment by jcs090218** [comment]
Can you fix these compile warnings? Thanks! :)

```
In lsp-csharp--omnisharp-metadata-uri-handler:
lsp-csharp.el:358:2: Warning: docstring has wrong usage of unescaped single quotes (use \= or different quoting)

In lsp-csharp--omnisharp-uri->path-fn:
lsp-csharp.el:394:2: Warning: docstring wider than 80 characters
```

File: clients/lsp-csharp.el [hunk 5]
```diff
@@ -334,6 +348,61 @@ using the `textDocument/references' request."
...
+        (insert source)))
+
+    file-location))
+
+(defun lsp-csharp--omnisharp-uri->path-fn (uri)
+  "Custom implementation of lsp--uri-to-path function to glue omnisharp's metadata uri."
+  (if (string-match-p lsp-csharp--omnisharp-metadata-uri-re uri)
+      (lsp-csharp--omnisharp-metadata-uri-handler uri)
+    (lsp--uri-to-path-1 uri))
+  )
```
Comment (jcs090218) [inline clients/lsp-csharp.el:399]: Let place this `)` to the previous line.

**Comment by cheerio-pixel** [comment]
@jcs090218 Anything more?

**Review by jcs090218 — APPROVED** [review]
LGTM! Thank you! :)

**Pull request merged**